extender.go

```diff
@@ -113,38 +121,272 @@ func (f *ExtenderFactory) verifyAPIs(groupExt []buildpack.GroupElement, logger l
+	// save sparse image (manifest and config)
+	imageHash, err := image.Digest()
+	if err != nil {
+		return fmt.Errorf("getting image hash: %w", err)
+	}
+	toPath := filepath.Join(e.ExtendedDir, "run", imageHash.String())
+	layoutPath, err := selective.Write(toPath, empty.Index) // FIXME: this should use the imgutil layout/sparse package instead, but for some reason sparse.NewImage().Save() fails when the provided base image is already sparse
+	if err != nil {
+		return fmt.Errorf("initializing selective image: %w", err)
+	}
+	if err = layoutPath.AppendImage(image); err != nil {
+		return fmt.Errorf("saving selective image: %w", err)
+	}
+	// get all image layers (we will only copy those following the original top layer)
+	layers, err := image.Layers()
+	if err != nil {
+		return fmt.Errorf("getting image layers: %w", err)
+	}
+	var (
+		currentHash  v1.Hash
+		needsCopying bool
+	)
+	if origTopLayerHash == "" { // if the original base image had no layers, copy all the layers
+		needsCopying = true
+	}
+	group, _ := errgroup.WithContext(context.TODO())
+	for _, currentLayer := range layers {
+		currentHash, err = currentLayer.Digest()
 		if err != nil {
-			return err
+			return fmt.Errorf("getting layer hash: %w", err)
 		}
-		if buildDockerfile != nil {
-			e.Logger.Debugf("Found build Dockerfile for extension '%s'", ext.ID)
-			dockerfiles = append(dockerfiles, *buildDockerfile)
+		switch {
+		case needsCopying:
+			currentLayer := currentLayer // allow use in closure
+			group.Go(func() error {
+				return copyLayer(currentLayer, toPath)
+			})
+		case currentHash.String() == origTopLayerHash:
+			needsCopying = true
+			continue
+		default:
+			continue
 		}
 	}
-	options := extend.Options{
-		CacheTTL:    e.CacheTTL,
-		IgnorePaths: []string{e.AppDir, e.LayersDir, e.PlatformDir},
+	return group.Wait()
+}
+
+func copyLayer(layer v1.Layer, toSparseImage string) error {
+	digest, err := layer.Digest()
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(filepath.Join(toSparseImage, "blobs", digest.Algorithm, digest.Hex))
+	if err != nil {
+		return err
 	}
-	digest, err := name.NewDigest(e.ImageRef)
+	defer f.Close()
+	rc, err := layer.Compressed() // FIXME: if exporting to a daemon, this should be uncompressed
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	_, err = io.Copy(f, rc)
+	return err
+}
+
+const (
+	argBuildID = "build_id"
+	argUserID  = "user_id"
+	argGroupID = "group_id"
+)
+
+func (e *Extender) extend(kind string, baseImage v1.Image, logger log.Logger) (v1.Image, error) {
+	logger.Debugf("Extending base image for %s: %s", kind, e.ImageRef)
+	dockerfiles, err := e.dockerfilesFor(kind, logger)
 	if err != nil {
-		return fmt.Errorf("failed to get digest for reference %s: %w", e.ImageRef, err)
+		return nil, fmt.Errorf("getting %s dockerfiles: %w", kind, err)
+	}
+
+	var (
+		configFile *v1.ConfigFile
+		rebasable  = true // for now, we don't require the initial base image to have io.buildpacks.rebasable=true
```

Comment:
that's gonna be a tough one right? bc if we don't start that way, then technically we'd break backwards compatibility to ever require it?   is the issue that basically nobody's gonna decorate their images like that,  so we kind of just say "the absence of rebasable=false label is true"?  

I guess i'm questioning whether the comment should imply that in the future we'll really have a default of rebaseable false,  given that we're starting with a default of rebaseable true.   are we skirting a spec change by saying "yeah yeah, we'll do it later"?? 😆

Reply:
Fair point - I'l remove "for now" because I don't think we'll ever actually make this a requirement.